File: 0377-combination-sum-iv/0377-combination-sum-iv.py
class Solution:
    
    def util(self, nums: List[int], target: int, n: int) -> int:
        if not target:
            return 1
        if target<0:
            return 0
        if self.dp[target]!=-1:
            return self.dp[target]
        ans = 0
        for i in nums:
            ans = ans +  self.util(nums, target-i, n)
        self.dp[target]= ans
        return ans
    
    def combinationSum4(self, nums: List[int], target: int) -> int:
        self.dp = [-1 for _ in range(1001)]
        n = len(nums)
        return self.util(nums, target, n)
        
        

File: 0377-combination-sum-iv/0377-combination-sum-iv.cpp
class Solution {
public:
    int dp[1001];
    
    int util(vector<int>& nums, int n, int target) {
        if(target==0)
            return 1;
        if(target<0)
            return 0;
        if(dp[target]!=-1)
            return dp[target];
        int ans =0;
        for(int i=0;i<n;i++) {
            ans += util(nums, n, target-nums[i]);
        }
        return dp[target]=ans;
    }
    
    
    int combinationSum4(vector<int>& nums, int target) {
        int n=  nums.size();
        memset(dp, -1, sizeof(dp));
        return util(nums, n, target);
    }
};